arci-ros2/src/navigation.rs

```diff
@@ -63,59 +63,91 @@ impl Navigation for Ros2Navigation {
+                    tokio::spawn(async move {
+                        let mut clock = r2r::Clock::create(r2r::ClockType::RosTime).unwrap();
+                        let now = clock.get_now().unwrap();
+                        let goal = NavigateToPose::Goal {
+                            pose: msg::PoseStamped {
+                                header: Header {
+                                    frame_id: frame_id.to_string(),
+                                    stamp: Time {
+                                        sec: now.as_secs() as i32,
+                                        nanosec: now.subsec_nanos(),
+                                    },
+                                },
+                                pose: to_pose_msg(&goal),
+                            },
+                            ..Default::default()
+                        };
 
-        let has_reached_set = has_reached.clone();
-        let result_cb = Box::new(move |r: NavigateToPose::Result| {
-            info!("final result {:?}", r);
-            *has_reached_set.lock().unwrap() = Some(r);
+                        is_available.await.unwrap();
+                        let send_goal_request = action_client.send_goal_request(goal).unwrap();
+                        let (goal, result, feedback) = send_goal_request.await.unwrap();
+                        *current_goal_clone.lock().unwrap() = Some(goal.clone());
+                        tokio::spawn(
+                            async move { feedback.for_each(|_| std::future::ready(())).await },
+                        );
+                        result.await.unwrap(); // TODO: handle goal state
+                        is_done.store(true, Ordering::Relaxed);
+                    });
+                    loop {
+                        node.lock()
+                            .unwrap()
+                            .spin_once(std::time::Duration::from_millis(100));
+                        tokio::task::yield_now().await;
+                        if is_done_clone.load(Ordering::Relaxed) {
+                            break;
+                        }
+                    }
+                    *current_goal.lock().unwrap() = None;
+                    let _ = sender.send(());
+                });
         });
-        self.action_client
-            .send_goal_request(goal, cb, feedback_cb, result_cb)
-            .unwrap();
-        let spin_node = self.node.clone();
-        let start_time = std::time::Instant::now();
         let wait = WaitFuture::new(async move {
-            const SLEEP_DURATION: std::time::Duration = std::time::Duration::from_micros(100);
-            while has_reached.lock().unwrap().is_none() && start_time.elapsed() < timeout {
-                spin_node.lock().unwrap().spin_once(SLEEP_DURATION);
-                std::thread::sleep(SLEEP_DURATION);
-            }
-            // TODO: Handle the result and timeout
+            let _ = receiver.await;
```

Comment:
Implementing the timeout is difficult?
If you skip it, please add a TODO comment for this function.

Reply:
timeout is implemented in https://github.com/openrr/openrr/pull/501/commits/a12be58875ccc7107a889fc1046dc9ad7eafef9a